bpf/cgroup/bpf_cgtracker.c

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Tetragon */
+
+#include "vmlinux.h"
+#include "api.h"
+#include "bpf_helpers.h"
+#include "bpf_cgroup.h"
+#include "bpf_tracing.h"
+#include "cgtracker.h"
+
+char _license[] __attribute__((section(("license")), used)) = "GPL";
+#ifdef VMLINUX_KERNEL_VERSION
+int _version __attribute__((section(("version")), used)) =
+	VMLINUX_KERNEL_VERSION;
+#endif
+
+/* new kernel cgroup definition */
+struct cgroup___new {
+	int level;
+	struct cgroup *ancestors[];
+} __attribute__((preserve_access_index));
+
+FUNC_INLINE __u64 cgroup_get_parent_id(struct cgroup *cgrp)
+{
+	struct cgroup___new *cgrp_new = (struct cgroup___new *)cgrp;
+
+	// for newer kernels, we can access use ->ancestors to retrieve the parent
+	if (bpf_core_field_exists(cgrp_new->ancestors)) {
+		int level = get_cgroup_level(cgrp);
+
+		if (level <= 0)
+			return 0;
+		return BPF_CORE_READ(cgrp_new, ancestors[level - 1], kn, id);
+	}
+
+	// otherwise, go over the parent pointer
+	struct cgroup_subsys_state *parent_css = BPF_CORE_READ(cgrp, self.parent);
+
+	if (parent_css) {
+		struct cgroup *parent = container_of(parent_css, struct cgroup, self);
+		__u64 parent_cgid = get_cgroup_id(parent);
+		return parent_cgid;
+	}
+
+	return 0;
+}
+
+__attribute__((section(("raw_tracepoint/cgroup_mkdir")), used)) int
+tg_cgtracker_cgroup_mkdir(struct bpf_raw_tracepoint_args *ctx)
+{
+	struct cgroup *cgrp;
+	__u64 cgid, cgid_parent, *cgid_tracker;
+
+	cgrp = (struct cgroup *)ctx->args[0];
+	cgid = get_cgroup_id(cgrp);
+	if (cgid == 0)
+		return 0;
+	cgid_parent = cgroup_get_parent_id(cgrp);
```

Comment:
Hmm didn't have time to review all, but want to point this is not cgroupv1 compatible, the mkdir and release delete will work on any cgoupv1 hierarchies, and if kernfs ids clash could lead to corruption?

The old branch has this check: https://github.com/cilium/tetragon/blob/pr/tixxdz/cgroup-bpf-full/bpf/cgroup/bpf_cgroup_mkdir.c#L33 

So this could break in cgroupv1. Will check closely later.

Reply:
I think it's fine to have this only for group v2 at the moment. That being said, it's unclear what the problem is for v1. What are the kernfs IDs that could clash?

Reply:
you are following all hierarchies in cgroupv1! including the ones that we don't track, why?

The kernfs IDs are cgroup IDs if they are re-used you remove them from the tracking and it won't work anymore, see your cgrp release bpf program.

Reply:
> you are following all hierarchies in cgroupv1! including the ones that we don't track, why?

We can add a filter to tack only the configured hierarchy for v1. My first goal was to support cgroup v2 so that's what the first implementation does. Any cgroup v1 support might or might not work since we do not have tests for it.

> The kernfs IDs are cgroup IDs if they are re-used you remove them from the tracking and it won't work anymore, see your cgrp release bpf program.

Not sure I understand. Are you saying that cgroup ids are unique only within a specific hierarchy?

Reply:
To make sure I understand correctly, are you saying that it's possible for two cgroup nodes of different hierarchies to have the same kernfs id?